middleware.go

```diff
@@ -73,6 +89,66 @@ func EchoFirstTraceNodeInfo(propagator propagation.TextMapPropagator) func(http.
 	}
 }
 
+// findWRPTraceHeaders looks for trace headers in WRP messages
+// Returns a context propagated with the trace information
+func findWRPTraceHeaders(ctx context.Context, r *http.Request) context.Context {
+
+	// Check for trace info in Xmidt headers
+	var traceHeaders []string
+	xmidtHeaders := r.Header.Values("X-midt-Headers")
+	if len(xmidtHeaders) != 0 {
+		traceHeaders = xmidtHeaders
+	} else {
+
+		// Check for trace info in the request body
+		reader := r.Body
+		bodyData, err := io.ReadAll(reader)
+		if err != nil {
+			log.Printf("failed to read body: %s", err)
+		}
+
+		// Reset request body
+		newBody := bytes.NewReader(bodyData)
+		r.Body = io.NopCloser(newBody)
+		r.GetBody = func() (io.ReadCloser, error) {
+			_, err := newBody.Seek(0, 0)
+			return io.NopCloser(newBody), err
+		}
+
+		var msg wrp.Message
+		if err := json.Unmarshal(bodyData, &msg); err != nil {
+			log.Printf("failed to unmarshal payload: %s", err)
+		}
+
+		if r.Header.Get("Content-Type") == "application/msgpack" {
```

Comment:
The knowledge of what content types are associated with which `wrp` formats more properly belongs in `wrp-go`.  We have some of that functionality already: https://pkg.go.dev/github.com/xmidt-org/wrp-go@v1.3.4/wrp#FormatFromContentType

What's missing is a way to fully extract a `wrp` message from a request, using the `context` as a "cache".  I've added the following issues to `wrp-go` to address this:

- https://github.com/xmidt-org/wrp-go/issues/114
- https://github.com/xmidt-org/wrp-go/issues/115

Reply:
Updated in https://github.com/xmidt-org/wrp-go/pull/116/